libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/PredictiveCache.kt

```diff
@@ -7,17 +7,17 @@ import com.mapbox.navigator.PredictiveCacheController
 
 object PredictiveCache {
 
-    private val cachedNavigationPredictiveCacheControllers =
-        mutableListOf<PredictiveCacheController>()
-    private val cachedMapsPredictiveCacheControllers =
+    internal val cachedNavigationPredictiveCacheControllers =
+        mutableSetOf<PredictiveCacheController>()
+    internal val cachedMapsPredictiveCacheControllers =
         mutableMapOf<Any, MutableMap<String, PredictiveCacheController>>()
 
-    private val navPredictiveCacheLocationOptions =
-        mutableListOf<PredictiveCacheLocationOptions>()
-    private val mapsPredictiveCacheLocationOptions =
+    internal val navPredictiveCacheLocationOptions =
+        mutableSetOf<PredictiveCacheLocationOptions>()
+    internal val mapsPredictiveCacheLocationOptions =
```

Comment:
This looks like a hack. We still try to concurrently modify the list, it just turns out that we are not adding any elements to it because all are already there. If that wasn't the case, we would see a crash again. This means that the loop itself in `init` is not built correctly in the first place. The loop shouldn't try adding new controllers to the list, they all are already there, it should only create a new native controller.

Reply:
Clarifying - I'm not against using sets here, I just think we should refactor the loop as well.

Reply:
there are no controllers, we clear them
```kotlin
 MapboxNativeNavigatorImpl.setNativeNavigatorRecreationObserver {
            cachedNavigationPredictiveCacheControllers.clear()
            cachedMapsPredictiveCacheControllers.clear()
            
            ...
}
```

after that we iterate over cached options and recreate controllers

```kotlin
 MapboxNativeNavigatorImpl.setNativeNavigatorRecreationObserver {
          ...
          navPredictiveCacheLocationOptions.forEach {
                createNavigationController(it)
            }
          ...
}
```

but, `fun createNavigationController(options: PredictiveCacheLocationOptions)`
caches `options` , so we iterate over collection and modify it.
it's a valid operation, but lists throw `ConcurrentModificationException`. To use it with lists, we need `Iterator` instance. But `Set` allows it without an `Iterator` and as a bonus we get non-duplicates

Reply:
@LukasPaczos got you concern, will double check

Reply:
@LukasPaczos ready